Alejo Vinluan
abv210001
CS 4395.002

# WordNet Basics

## Overview
WordNet is a network and organizational tactic of all words including nouns, verbs, adjectives, and adverbs. For every word in the network, WordNet lists short definitions, synonyms, usage, and relations to other words.

In [18]:
import nltk
nltk.download('wordnet')

[nltk_data] Error loading wordnet: <urlopen error [SSL:
[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:
[nltk_data]     unable to get local issuer certificate (_ssl.c:1129)>


False

## Short Demonstration with a Noun
In this excercise, the noun "wizard" will be used

In [19]:
noun = "wizard"

All of the synsets for "wizard" will be:

In [20]:
synsets = nltk.corpus.wordnet.synsets(noun)
print("Synsets for `wizard`:")
for synset in synsets:
    print("\t-", synset.name().split('.')[0])

Synsets for `wizard`:
	- ace
	- sorcerer
	- charming


This chunk of code will select 1 synset from the list (at random). Then, the code will will extract the definition, usage examples, and lemmas from the sysnset. Finally, the code will traverse up the hierarchy of synsets, outputting them while traversing.

In [21]:
# Importing random to randomly select a synset from noun
import random

# Setting a seed to reproduce results for testing
random.seed(1234)

# Choose a synset
random_synset = random.choice(synsets)

# Print which synset was randomly selected
print("Random Synset derived from 'Wizard':", random_synset.name().split('.')[0], "\n")

# Print the definition of the randomly selected synset
print("Definition:", random_synset.definition(), "\n")

# Print usage examples if they exist
if len(random_synset.examples()) == 0:
    print("No examples found.\n")
else:
    print("Usage examples:")
    for example in random_synset.examples():
        print("\t- ", example)
        
# Print lemmas if they exist
if len(random_synset.lemmas()) == 0:
    print("No lemmas found.\n")
else:
    print("Lemmas:")
    for lemma in random_synset.lemmas():
        print("\t- ", lemma.name().split('.')[0])
    
# Traversing the hierarchy
# Find whether the current synset has a hypernym
if len(random_synset.hypernyms()) > 0:
    # Begin traversing the hierarchy
    print("\nTraversing up the hierarchy for:", random_synset.name().split('.')[0])
    # Select the first hypernym for the given synset
    hypernym = random_synset.hypernyms()[0]
    # Continue looping while hypernyms are available
    while hypernym:
        print("\t-", hypernym.name().split('.')[0])
        # Update the hypernym to the current synset's hypernym. Unless the hypernym is
        #  the top of the synset list, which is entity.
        if len(hypernym.hypernyms()) > 0 and hypernym.name() != "entity.n.01":
            hypernym = hypernym.hypernyms()[0]
        else:
            break
else:
    print('No hypernyms found for synset:', random_synset.name().split('.')[0])

Random Synset derived from 'Wizard': sorcerer 

Definition: one who practices magic or sorcery 

No examples found.

Lemmas:
	-  sorcerer
	-  magician
	-  wizard
	-  necromancer
	-  thaumaturge
	-  thaumaturgist

Traversing up the hierarchy for: sorcerer
	- occultist
	- person
	- causal_agent
	- physical_entity
	- entity


When traversing up the WordNet hierarchy, WordNet will go from specific to broad examples. Sorcerer's hypernym is 'occultist', which is defined as somebody who practices magic, astrology, alchemy, or other supernatural powers. This implies that all sorcerers are occultists, but not all occultists are sorcerers. The same can be said from occultists to person. A person is an incredibly broad category that can encapsulate all occultists, but is not necesarilly an occultist. Eventually, the hypernyms turn into an entity, which can represent anything.

From the randomly selected synset, here is a list of it's hypernyms, hyponyms, meronyms, holonyms, and antonyms if they're available. 

In [22]:
# Print hypernyms
if len(random_synset.hypernyms()) > 0:
    print("Hypernyms:")
    for hypernym in random_synset.hypernyms():
        print("\t-", hypernym.name().split('.')[0])
else:
    print("Found no hypernyms for", random_synset.name().split('.')[0], "\n")
    
# Print hyponyms
if len(random_synset.hyponyms()) > 0:
    print("Hyponyms:")
    for hyponym in random_synset.hyponyms():
        print("\t-", hyponym.name().split('.')[0])
else:
    print("Found no hyponyms for", random_synset.name().split('.')[0], "\n")
    
# Print meronyms
if len(random_synset.part_meronyms()) > 0:
    print("Meronyms:")
    for meronym in random_synset.part_meronyms():
        print("\t-", meronym.name().split('.')[0])
else:
    print("\nFound no meronyms for:", random_synset.name().split('.')[0], "\n")
    
# Print antonyms
if len(random_synset.lemmas()) > 0 and len(random_synset.lemmas()[0].antonyms()) > 0:
    print("Antonyms:")
    for antonym in random_synset.lemmas()[0].antonyms():
        print("\t-", antonym.name().split('.')[0])
else:
    print("Found no antonyms for:", random_synset.name().split('.')[0], "\n")

Hypernyms:
	- occultist
Hyponyms:
	- enchanter
	- exorcist
	- magus
	- sorceress
	- witch_doctor

Found no meronyms for: sorcerer 

Found no antonyms for: sorcerer 



## Short Demonstration with a Verb
The verb "hypnotize" will be used. 

The following code chunk will output all synsets for "bewitch".

In [23]:
verb = "bewitch"
synsets = nltk.corpus.wordnet.synsets(verb)
print("Synsets for `bewitch`: ")
for synset in synsets:
    print("\t-", synset.name().split('.')[0])

Synsets for `bewitch`: 
	- capture
	- magnetize
	- hex


From "bewitch", a random synset will be chosen. The definition, usage examples, and lemmas will be printed. Finally, the code will traverse up synset hierarchy and output at each level of the hierarchy.

In [24]:
# Choose a synset
random_synset = random.choice(synsets)

# Print which synset was randomly selected
print("Random Synset derived from 'Bewitch':", random_synset.name().split('.')[0], "\n")

# Print the definition of the randomly selected synset
print("Definition:", random_synset.definition(), "\n")

# Print usage examples if they exist
if len(random_synset.examples()) == 0:
    print("No examples found.\n")
else:
    print("Usage examples:")
    for example in random_synset.examples():
        print("\t- ", example)
        
# Print lemmas if they exist
if len(random_synset.lemmas()) == 0:
    print("No lemmas found.\n")
else:
    print("Lemmas:")
    for lemma in random_synset.lemmas():
        print("\t- ", lemma.name().split('.')[0])
    
# Traversing the hierarchy
# Find whether the current synset has a hypernym
if len(random_synset.hypernyms()) > 0:
    # Begin traversing the hierarchy
    print("\nTraversing up the hierarchy for:", random_synset.name().split('.')[0])
    # Select the first hypernym for the given synset
    hypernym = random_synset.hypernyms()[0]
    # Continue looping while hypernyms are available
    while hypernym:
        print("\t-", hypernym.name().split('.')[0])
        # Update the hypernym to the current synset's hypernym. Unless the hypernym is
        #  the top of the synset list, which is entity.
        if len(hypernym.hypernyms()) > 0:
            hypernym = hypernym.hypernyms()[0]
        else:
            break
else:
    print('No hypernyms found for synset:', random_synset.name().split('.')[0])

Random Synset derived from 'Bewitch': capture 

Definition: attract; cause to be enamored 

Usage examples:
	-  She captured all the men's hearts
Lemmas:
	-  capture
	-  enamour
	-  trance
	-  catch
	-  becharm
	-  enamor
	-  captivate
	-  beguile
	-  charm
	-  fascinate
	-  bewitch
	-  entrance
	-  enchant

Traversing up the hierarchy for: capture
	- attract


It seems like the verb hierarchy continues to present verbs that are broader than the currently selected verb. However, there is no top level for these verbs like "entity" is for nouns. The verbs that are broader may not also mean the same as the more specific verb. For example, attract and capture are not interchangeable. Although capture can mean attract in some ways.

## Morphy Demonstration
The following code block will find different forms of the verb 'casting'.

In [25]:
from nltk.corpus import wordnet
print(wordnet.morphy('casting',wordnet.VERB))

cast


In this instance there was only one word Morphy found for 'casting': cast. 

## Wu-Palmer Simlarity Metric and Lesk Algorithm
The following code block will utilize the Wu-Palmer Similarity Metric and Lesk Algorithm to find how similar the words 'Sorcerer' and 'Alchemist' are.

In [34]:
sorcerer = nltk.corpus.wordnet.synsets('sorcerer')[0]
alchemist = nltk.corpus.wordnet.synsets('alchemist')[0]

wup_score = wordnet.synset(sorcerer.name()).wup_similarity(wordnet.synset(alchemist.name()))
print(wup_score)

0.6666666666666666


In [53]:
from nltk.wsd import lesk

sample_sentence = ['The', 'sorcerer', 'casted', 'a', 'spell', '.']
print(lesk(sample_sentence, 'sorcerer', 'n'))

sample_sentence = ['The', 'alchemist', 'brewed', 'a', 'potion', '.']
print(lesk(sample_sentence, 'alchemist', "n"))

sample_sentence = ['The', 'sorcerer', 'casted', 'a', 'spell', 'against', 'the', 'alchemist', '.']
print(lesk(sample_sentence, 'sorcerer', 'n'))
print(lesk(sample_sentence, 'alchemist', 'n'))


Synset('sorcerer.n.01')
Synset('alchemist.n.01')
Synset('sorcerer.n.01')
Synset('alchemist.n.01')


The Wu-Palmer Similarity Metric finds that 'sorcerer' and 'alchemist' have a 66% similarity metric. However, after using Lesk, there were no similarities found. This may be because there is no ambiguity between 'sorcerer' and 'algorithm'. A better metric would be using the word 'bank' since that could be used to differentiate between a bank account, a river bank, or a slope bank.

## SentiWordNet
SentiWordNet is built on WordNet and can give words a score of positive, negative, or objective. This can be useful since a program can breakdown the sentiment of sentences or pages. For example, when scraping the Wikipedia page for "Disneyland", the program can find a positive sentiment towards Disneyland and understand that Disneyland is generally seen as positive.

The following code will utilize the emotionally charged word 'fear', find it's senti-synsets, and output the polarity scores. 

In [59]:
from nltk.corpus import sentiwordnet as swn

senti_synset = swn.senti_synset('fear.v.01')
print(senti_synset)

print("Positive Score:", senti_synset.pos_score())
print("Negative Score:", senti_synset.neg_score())
print("Objective Score:", senti_synset.obj_score())


Positive Score: 0.125
Negative Score: 0.625
Objective Score: 0.25


The next example will find the sentiment score of the sentence, "I absolutely love getting scared by Haunted Houses."

In [65]:
sentence = 'I absolutely love getting scared by Haunted Houses'
positive_score = 0
negative_score = 0

tokens = sentence.split()
for token in tokens:
    syn_list = list(swn.senti_synsets(token))
    if syn_list:
        syn = syn_list[0]
        negative_score += syn.neg_score()
        positive_score += syn.pos_score()

print("Positive Score:", positive_score)
print("Negative Score:", negative_score)

<iodine.n.01: PosScore=0.0 NegScore=0.0>
<absolutely.r.01: PosScore=0.5 NegScore=0.0>
<love.n.01: PosScore=0.625 NegScore=0.0>
<acquiring.n.01: PosScore=0.0 NegScore=0.0>
<frighten.v.01: PosScore=0.375 NegScore=0.0>
<by.r.01: PosScore=0.0 NegScore=0.0>
<haunt.v.01: PosScore=0.0 NegScore=0.0>
<house.n.01: PosScore=0.0 NegScore=0.0>
Positive Score: 1.5
Negative Score: 0.0


I am surprised that there is a 0 in negative score after utilizing the word "scared", which turned into a Synset of "frightened". I suspect that WordNet may have utilized the context of the sentence as a whole to determine the sentiment of the entire sentence. Knowing these scores is important to breakdown how to handle and process certain emotionally charged statements within Natural Language Processing.

## Collocations
Collacations are words that are generally put together, but do not make literal sense. They are together often enough that the meaning is understood, but either word cannot be replaced by a synonym to get the same meaning. Some examples of collocations include "heavy rain" and "high school".

The following code block will output collocations for The Inaugural Corpus. Then, it will select one coallocation, and calculate mutual information.

In [86]:
import re
import math
from nltk.book import text4

# Print collocation list
print("Collocation List:", text4.collocation_list(), "\n")

# Select a random collocation from the list
random.seed(123456)
random_collocation = random.choice(text4.collocation_list())
print("Random Collocation:", random_collocation, "\n")

# Count total occurences of "Chief Justice"
chief_count = 0
justice_count = 0
chief_justice_count = 0

for word in text4:
    if word.lower() == 'chief':
        chief_count += 1
    if word.lower() == 'justice':
        justice_count += 1

chief_justice_count = len(re.findall("Chief Justice", ' '.join(text4)))
total_length = len(text4)

# Calculate PMI (Point-Wise Mutual Information)
pmi = math.log2((chief_justice_count / total_length) / ((chief_count / total_length) * (justice_count / total_length)))
print("PMI Score:", pmi)

Collocation List: [('United', 'States'), ('fellow', 'citizens'), ('years', 'ago'), ('four', 'years'), ('Federal', 'Government'), ('General', 'Government'), ('American', 'people'), ('Vice', 'President'), ('God', 'bless'), ('Chief', 'Justice'), ('one', 'another'), ('fellow', 'Americans'), ('Old', 'World'), ('Almighty', 'God'), ('Fellow', 'citizens'), ('Chief', 'Magistrate'), ('every', 'citizen'), ('Indian', 'tribes'), ('public', 'debt'), ('foreign', 'nations')] 

Random Collocation: ('Chief', 'Justice') 

PMI: 8.670392753575022


The phrase "Chief Justice" has a PMI score of 8.67 which suggests that it is a strong coallocation. The word "Chief" is defined as a leader of a clan and the word "Justice" is defined as a judge or mastrate. The leader of a clan of judges generally does not make sense in the English language. However, "Chief Justice" as a position is well known and understandable.